soroban-sdk/src/cost_estimate.rs

```diff
@@ -0,0 +1,111 @@
+use soroban_env_host::{fees::FeeConfiguration, FeeEstimate, InvocationResources};
+
+use crate::{testutils::budget::Budget, Env};
+
+pub struct CostEstimate {
+    env: Env,
+}
+
+impl CostEstimate {
+    pub(crate) fn new(env: Env) -> Self {
+        Self { env }
+    }
+
+    /// Enables detailed per-invocation resource cost metering.
+    ///
+    /// The top-level contract invocations and lifecycle operations (such as
+    /// `register` or `env.deployer()` operations) will be metered and
+    /// all the metering information will reset in-between them. Metering will
+    /// not be reset while inside the call (e.g. if a contract calls or creates
+    /// another contract, that won't reset metering).
+    ///
+    /// The metered resources for the last invocation can be retrieved with
+    /// `resources()`, the estimated fee corresponding to these resources can be
+    /// retrieved with `fee()`, and the detailed CPU and memory memory
+    /// breakdown can be retrieved with `detailed_metering()`.
+    ///
+    /// While the resource metering may be useful for contract optimization,
+    /// keep in mind that resource and fee estimation may be imprecise. Use
+    /// simulation with RPC in order to get the exact resources for submitting
+    /// the transactions to the network.    
+    pub fn enable(&self) {
```

Comment:
We should make this function pub(crate) since it's intended use is via the EnvTestConfig.

Reply:
I've tried that, however the issue is that we allow updating the config without resetting the env via `set_config`. The host function we now is one way (I didn't anticipate this ever being disabled after being enabled). So we can't provide appropriate config behavior right now. I see several options here:

- Leave things as is in this PR (i.e. enable via `cost_estimate`) directly
- Updating the host to also allow disabling this. This might require some semver wrestling though, as this is a public function.
- Maybe just enable this by default after all. The scope of potential breakage seems really narrow (observing the budget and asserting on exact values). We also don't intentionally try to avoid this sort of breakages in general, as any internal setup change might lead to the changes in the observed budget. I realize that here the change is more explicit than e.g. updating the implementation of instantiation function, but likely the benefits (better usability, less code, less cleanup after next release) outweigh the (improbable) breakage risks.

Reply:
Hmm. I think we can do this:

If set_config is called with turning the value off, when the value is already turned on (should be an easy comparison), panic with an informative message saying that once it is enabled it cannot be disabled.

This addresses the 90% case, people will turn it on, and leave it on.

And a panic here, is fine.